text/0000-let-expression.md

```diff
@@ -0,0 +1,917 @@
+dozens of them just in rust-clippy. Note that if-let-chain alone can't solve this problem, because you have access to
+variables of if-let inside of block, but you have access to variables of a binding statement under it, thus you can
+save one indentation.
+
+`||` is not only useful for let-else style things. This is a real example from [deno](https://github.com/denoland/deno):
+
+```rust
+let nread = if let Some(s) = resource.downcast_rc::<ChildStdoutResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<ChildStderrResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<TcpStreamResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<TlsStreamResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<UnixStreamResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<StdFileResource>() {
+    s.read(buf).await?
+} else {
+    return Err(not_supported());
+};
+```
+Which by `||` counterpart of if-let-chain can become:
+```rust
+let nread = if let Some(s) = resource.downcast_rc::<ChildStdoutResource>()
+    || let Some(s) = resource.downcast_rc::<ChildStderrResource>()
+    || let Some(s) = resource.downcast_rc::<TcpStreamResource>()
+    || let Some(s) = resource.downcast_rc::<TlsStreamResource>()
+    || let Some(s) = resource.downcast_rc::<UnixStreamResource>()
+    || let Some(s) = resource.downcast_rc::<StdFileResource>() {
+    s.read(buf).await?
+} else {
+    return Err(not_supported());
+};
+```
+This is smaller and doesn't repeat a code.
+
+A different class of practical usages of this RFC is let expression usage as a bool. People
+wrap their let expressions with `if expr { true } else { false }` manually. This need is almost met
+with `matches!` macro, but `if let true else false` is still a thing in rust code bases. Again from rust-clippy:
+```rust
+fn is_repeat_zero(&self, expr: &Expr<'_>) -> bool {
+    if_chain! {
+        if let ExprKind::Call(fn_expr, [repeat_arg]) = expr.kind;
+        if is_expr_path_def_path(self.cx, fn_expr, &paths::ITER_REPEAT);
+        if let ExprKind::Lit(ref lit) = repeat_arg.kind;
+        if let LitKind::Int(0, _) = lit.node;
+
+        then {
+            true
+        } else {
+            false
+        }
+    }
+}
+```
+After this RFC, we can write this:
+```rust
+fn is_repeat_zero(&self, expr: &Expr<'_>) -> bool {
+    let ExprKind::Call(fn_expr, [repeat_arg]) = expr.kind
+    && is_expr_path_def_path(self.cx, fn_expr, &paths::ITER_REPEAT)
+    && let ExprKind::Lit(ref lit) = repeat_arg.kind
+    && matches!(lit.node, LitKind::Int(0, _)) // you can use let expression here as well
+}
+```
+Some people may argue that current state is more readable, but [this lint](https://rust-lang.github.io/rust-clippy/master/index.html#needless_bool)
+is not agree with them.
```

Comment:
Umm, that lint only looks for `if <expr> { <bool> } else { <bool> }` so it has nothing to do with `if let`.

Reply:
If let expressions become a thing, it would. Certainly it can't suggest let expressions today. There is a lint for `if let pat = expr { true } else { false }` that suggests `matches!(expr, pat)` which is equivalent to suggesting this lint for let expressions. Also I added another example in this category.